File: android/app/src/main/kotlin/com/app/deenly/deenly/tajweed/TajweedChannelHandler.kt
package com.rnr.deenfocus.tajweed

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * MethodChannel + EventChannel wiring for Tajweed (ADR-006).
 * Port of ios/Runner/Tajweed/TajweedChannelHandler.swift — same channel names,
 * same method names, same JSON/error shape; only the native engine underneath differs.
 */
class TajweedChannelHandler(private val context: Context) : EventChannel.StreamHandler {
    /** Lazy — constructing [TajweedEngine] at FlutterEngine configure hung launch when
     * the engine eagerly parsed the 12MB canonical lexicon. */
    private val engine by lazy { TajweedEngine.getInstance(context) }
    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null
    private var didBindEngineEvents = false

    fun attach(methodChannel: MethodChannel, eventChannel: EventChannel) {
        eventChannel.setStreamHandler(this)
        methodChannel.setMethodCallHandler { call, result -> handle(call, result) }
    }

    private fun bindEngineEventsIfNeeded() {
        if (didBindEngineEvents) return
        didBindEngineEvents = true
        engine.onEvent = { payload -> eventSink?.success(payload) }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        // Do not construct TajweedEngine here — listen alone must stay cheap
        // at launch / Settings; engine is created on the first method that needs it.
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    fun onMemoryTrim() {
        if (didBindEngineEvents) engine.onMemoryWarning()
    }

    fun onAppBackground() {
        if (didBindEngineEvents) engine.onAppBackground()
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            // Flag get/set must NOT construct TajweedEngine (Settings loads this at launch).
            "getCanonicalLexicalProductionEnabled" -> {
                CanonicalLexicalAuthority.init(context)
                mainHandler.post {
                    result.success(
                        mapOf(
                            "enabled" to CanonicalLexicalAuthority.productionEnabled,
                            "defaultsKey" to CanonicalLexicalAuthority.KEY,
                        ),
                    )
                }
            }
            "setCanonicalLexicalProductionEnabled" -> {
                CanonicalLexicalAuthority.init(context)
                val enabled = CanonicalLexicalAuthority.parseEnabledArgument(call.argument("enabled"))
                    ?: CanonicalLexicalAuthority.parseEnabledArgument(
                        (call.arguments as? Map<*, *>)?.get("enabled"),
                    )
                if (enabled == null) {
                    mainHandler.post {
                        result.error(
                            TajweedErrorCode.INVALID_ARGS,
                            "enabled: Bool required (got ${call.argument<Any>("enabled")})",
                            null,
                        )
                    }
                    return
                }
                CanonicalLexicalAuthority.productionEnabled = enabled
                mainHandler.post {
                    result.success(
                        mapOf(
                            "enabled" to CanonicalLexicalAuthority.productionEnabled,
                            "defaultsKey" to CanonicalLexicalAuthority.KEY,
                        ),
                    )
                }
            }
            "isAvailable" -> {
                bindEngineEventsIfNeeded()
                result.success(engine.isAvailable())
            }
            "getRecordingState" -> {
                bindEngineEventsIfNeeded()
                result.success(engine.getRecordingState())
            }
            "ensureModel" -> {
                bindEngineEventsIfNeeded()
                engine.ensureModel { outcome -> deliver(outcome, result) }
            }
            "prepareModel" -> {
                bindEngineEventsIfNeeded()
                engine.prepareModel { outcome -> deliver(outcome, result) }
            }
            "startRecording" -> {
                bindEngineEventsIfNeeded()
                val surah = call.argument<Int>("surah")
                val ayah = call.argument<Int>("ayah")
                val expected = call.argument<String>("expectedArabic")
                val lexicalRef = call.argument<String>("lexicalReferenceArabic")
                if (surah == null || ayah == null || expected == null) {
                    mainHandler.post {
                        result.error(TajweedErrorCode.INVALID_ARGS, "surah, ayah, expectedArabic required", null)
                    }
                    return
                }
                engine.startRecording(
                    surah,
                    ayah,
                    expected,
                    lexicalReferenceArabic = lexicalRef,
                ) { outcome -> deliver(outcome, result) }
            }
            "stopRecordingAndScore" -> {
                bindEngineEventsIfNeeded()
                engine.stopRecordingAndScore { outcome -> deliverJson(outcome, result) }
            }
            "cancelRecording" -> {
                bindEngineEventsIfNeeded()
                engine.cancelRecording { outcome -> deliver(outcome, result) }
            }
            "dispose" -> {
                if (didBindEngineEvents) engine.dispose()
                mainHandler.post { result.success(null) }
            }
            "getActiveCoreMlInfo" -> {
                val store = ModelStore(context.applicationContext)
                val info = mutableMapOf<String, Any?>(
                    "available" to store.isAvailable(),
                    "version" to store.activeManifestVersion(),
                    "manifestVersion" to store.activeManifestVersion(),
                    "encoderSha256" to store.activeEncoderSha256(),
                )
                mainHandler.post { result.success(info) }
            }
            else -> mainHandler.post { result.notImplemented() }
        }
    }

    /** Every native completion crosses back to the main thread exactly once here —
     * Flutter's platform channels require MethodChannel.Result to be invoked on the main thread. */
    private fun deliver(outcome: Result<Unit>, result: MethodChannel.Result) {
        mainHandler.post {
            outcome.fold(
                onSuccess = { result.success(null) },
                onFailure = { e -> result.error(flutterCode(e), e.message, null) },
            )
        }
    }

    private fun deliverJson(outcome: Result<Map<String, Any?>>, result: MethodChannel.Result) {
        mainHandler.post {
            outcome.fold(
                onSuccess = { json -> result.success(json) },
                onFailure = { e -> result.error(flutterCode(e), e.message, null) },
            )
        }
    }

    companion object {
        fun flutterCode(e: Throwable): String = (e as? TajweedNativeException)?.code ?: TajweedErrorCode.INFERENCE_FAILED
    }
}

File: ios/Runner/Tajweed/TajweedChannelHandler.swift
import Flutter
import Foundation

/// MethodChannel + EventChannel wiring for Tajweed (ADR-006).
final class TajweedChannelHandler: NSObject, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?

  private var didBindEngineEvents = false

  func register(messenger: FlutterBinaryMessenger) {
    let method = FlutterMethodChannel(
      name: "com.app.deenly.deenly/tajweed",
      binaryMessenger: messenger
    )
    let events = FlutterEventChannel(
      name: "com.app.deenly.deenly/tajweed_events",
      binaryMessenger: messenger
    )
    events.setStreamHandler(self)
    // Do not touch TajweedEngine.shared here — constructing it creates
    // AVAudioEngine and would run on the launch path before first Flutter frame.
    method.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
  }

  private func bindEngineEventsIfNeeded() {
    guard !didBindEngineEvents else { return }
    didBindEngineEvents = true
    TajweedEngine.shared.onEvent = { [weak self] payload in
      self?.eventSink?(payload)
    }
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    // Do not construct TajweedEngine here — practice VM may listen before
    // ensureModel; engine is created on the first method that needs it.
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    // Flag get/set must not construct TajweedEngine — Settings can call these
    // while IndexedStack builds the Settings tab during first home paint.
    switch call.method {
    case "getCanonicalLexicalProductionEnabled":
      DispatchQueue.main.async {
        result(
          [
            "enabled": CanonicalLexicalAuthority.productionEnabled,
            "defaultsKey": CanonicalLexicalAuthority.defaultsKey,
          ] as [String: Any]
        )
      }
      return
    case "setCanonicalLexicalProductionEnabled":
      guard
        let args = call.arguments as? [String: Any],
        let enabled = CanonicalLexicalAuthority.parseEnabledArgument(args["enabled"])
      else {
        DispatchQueue.main.async {
          result(
            FlutterError(
              code: TajweedErrorCode.invalidArgs,
              message: "enabled: Bool required (got \(String(describing: (call.arguments as? [String: Any])?["enabled"])))",
              details: nil
            )
          )
        }
        return
      }
      CanonicalLexicalAuthority.productionEnabled = enabled
      DispatchQueue.main.async {
        result(
          [
            "enabled": CanonicalLexicalAuthority.productionEnabled,
            "defaultsKey": CanonicalLexicalAuthority.defaultsKey,
          ] as [String: Any]
        )
      }
      return
    default:
      break
    }

    bindEngineEventsIfNeeded()
    switch call.method {
    case "isAvailable":
      result(TajweedEngine.shared.isAvailable())
    case "getRecordingState":
      result(TajweedEngine.shared.getRecordingState())
    case "ensureModel":
      TajweedEngine.shared.ensureModel { [weak self] outcome in
        self?.deliver(outcome, result: result)
      }
    case "prepareModel":
      TajweedEngine.shared.prepareModel { [weak self] outcome in
        self?.deliver(outcome, result: result)
      }
    case "startRecording":
      guard
        let args = call.arguments as? [String: Any],
        let surah = args["surah"] as? Int,
        let ayah = args["ayah"] as? Int,
        let expected = args["expectedArabic"] as? String
      else {
        DispatchQueue.main.async {
          result(
            FlutterError(
              code: TajweedErrorCode.invalidArgs,
              message: "surah, ayah, expectedArabic required",
              details: nil
            )
          )
        }
        return
      }
      let lexicalRef = args["lexicalReferenceArabic"] as? String
      TajweedEngine.shared.startRecording(
        surah: surah,
        ayah: ayah,
        expectedArabic: expected,
        lexicalReferenceArabic: lexicalRef
      ) { [weak self] outcome in
        self?.deliver(outcome, result: result)
      }
    case "stopRecordingAndScore":
      TajweedEngine.shared.stopRecordingAndScore { [weak self] outcome in
        self?.deliverJSON(outcome, result: result)
      }
    case "cancelRecording":
      TajweedEngine.shared.cancelRecording { [weak self] outcome in
        self?.deliver(outcome, result: result)
      }
    case "dispose":
      TajweedEngine.shared.dispose()
      DispatchQueue.main.async { result(nil) }
    case "getDevCoreMlSource":
      DispatchQueue.main.async {
        result(
          [
            "source": TajweedDevModelOverride.preferredSource.rawValue,
            "overrideAllowed": TajweedDevModelOverride.isAllowed,
          ] as [String: Any]
        )
      }
    case "setDevCoreMlSource":
      guard TajweedDevModelOverride.isAllowed else {
        DispatchQueue.main.async {
          result(
            FlutterError(
              code: TajweedErrorCode.unsupported,
              message:
                "CoreML debug override requires a Debug build (SWIFT DEBUG flag). "
                + "Do a full rebuild: flutter clean && flutter run (not Profile/Release).",
              details: nil
            )
          )
        }
        return
      }
      guard
        let args = call.arguments as? [String: Any],
        let raw = args["source"] as? String,
        let source = TajweedDevModelSource(rawValue: raw)
      else {
        DispatchQueue.main.async {
          result(
            FlutterError(
              code: TajweedErrorCode.invalidArgs,
              message: "source must be catalog|official|diy",
              details: nil
            )
          )
        }
        return
      }
      TajweedDevModelOverride.preferredSource = source
      TajweedDevModelOverride.markForceResync()
      TajweedEngine.shared.dispose()
      DispatchQueue.main.async { result(nil) }
    case "getActiveCoreMlInfo":
      var info: [String: Any] = [
        "available": ModelStore.shared.isAvailable(),
        "overrideAllowed": TajweedDevModelOverride.isAllowed,
        "devSource": TajweedDevModelOverride.preferredSource.rawValue,
        "packMatchesSelection": TajweedDevModelOverride.activePackMatchesSelection(
          manifest: {
            guard let data = try? Data(contentsOf: ModelStore.shared.manifestURL),
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else { return nil }
            return obj
          }()
        ),
      ]
      if let m = try? Data(contentsOf: ModelStore.shared.manifestURL),
         let obj = try? JSONSerialization.jsonObject(with: m) as? [String: Any]
      {
        info["version"] = obj["version"] as? String ?? ""
        info["encoderApi"] = obj["encoderApi"] as? String ?? "multifunction"
        info["encoder"] = obj["encoder"] as? String ?? ""
      }
      info["encoderSha256"] = ModelStore.shared.activeEncoderSha256()
      info["manifestVersion"] = ModelStore.shared.activeManifestVersion()
      switch ModelStore.shared.encoderApi() {
      case .multifunction(let buckets):
        info["resolvedApi"] = "multifunction"
        info["encoderBuckets"] = buckets
      case .singleFunctionFixedLength(let t):
        info["resolvedApi"] = "single_function_fixed"
        info["encoderFixedT"] = t
      }
      DispatchQueue.main.async { result(info) }
    default:
      DispatchQueue.main.async { result(FlutterMethodNotImplemented) }
    }
  }

  /// Every native completion crosses back to the main thread exactly once here —
  /// Flutter's platform channels require FlutterResult to be invoked on the main thread.
  private func deliver(
    _ outcome: Result<Void, TajweedNativeError>,
    result: @escaping FlutterResult
  ) {
    DispatchQueue.main.async {
      switch outcome {
      case .success: result(nil)
      case .failure(let e): result(Self.flutterError(e))
      }
    }
  }

  private func deliverJSON(
    _ outcome: Result<[String: Any], TajweedNativeError>,
    result: @escaping FlutterResult
  ) {
    DispatchQueue.main.async {
      switch outcome {
      case .success(let json): result(json)
      case .failure(let e): result(Self.flutterError(e))
      }
    }
  }

  static func flutterError(_ e: TajweedNativeError) -> FlutterError {
    FlutterError(code: e.code, message: e.message, details: nil)
  }
}
